File system mocking example doesn't work

### Describe the bug

The file system mocking example provided in the documentation doesn't work.

https://vitest.dev/guide/mocking.html#example-3

Here are the step I took to get it working:

1) There are 2 issues in the `readHelloWorld` implementation:
- the `path` param isn't used
- the file encoding is absent, which makes it return a `Buffer` instead of string

```js
import { readFileSync } from 'node:fs'

export function readHelloWorld(path) {
  return readFileSync('./hello-world.txt')
}
```

Modified version:

```js
import { readFileSync } from "node:fs";

export function readHelloWorld(path, encoding = "utf-8") {
  return readFileSync(path, { encoding });
}
```

2) The "should return correct text" test uses project-relative path, which also fails.

Changed it to:

```js
it("should return correct text", () => {
  vol.fromJSON({ ".": null }, "/tmp");

  const path = "/tmp/hello-world.txt";
  fs.writeFileSync(path, "hello world");

  const text = readHelloWorld(path);
  expect(text).toBe("hello world");
});
```

3) Renamed `__mocks__/fs.cjs` to `__mocks__/node:fs.cjs` as otherwise `vitest` did not pick up the mock.

After these 3 steps the tests started working.

However, if I rename `__mocks__/fs/promises.cjs` to `__mocks__/node:fs/promises.cjs`, then tests fail with the following error:

```
 FAIL  test/hello-world.test.js [ test/hello-world.test.js ]
Error: EISDIR: illegal operation on a directory, read <PROJECT_HOME>/__mocks__/node:fs
 ❯ loadAndTransform node_modules/vite/dist/node/chunks/dep-D8YhmIY-.js:52358:16
```

So is there a way to mock both sync and async `fs` methods at the same time?



### Reproduction

1) Open https://stackblitz.com/edit/vitest-v2-fs-mock-bug
2) Download the sandbox to run the test locally
3) `npm install` 
4) `npm run test:run`

Observe all tests fail.

```
 ❯ test/hello-world.test.js (2)
   × should return correct text
   × can return a value multiple times

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 2 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  test/hello-world.test.js > should return correct text
Error: ENOENT: no such file or directory, open '<PROJECT_HOME>/hello-world.txt'
 ❯ createError node_modules/memfs/src/node/util.ts:145:17
 ❯ Volume.openFile node_modules/memfs/src/volume.ts:689:38
 ❯ Volume.openBase node_modules/memfs/src/volume.ts:701:23
 ❯ Volume.writeFileBase node_modules/memfs/src/volume.ts:1040:17
 ❯ Volume.writeFileSync node_modules/memfs/src/volume.ts:1064:10
 ❯ test/hello-world.test.js:17:6
     15| it('should return correct text', () => {
     16|   const path = './hello-world.txt';
     17|   fs.writeFileSync(path, 'hello world');
       |      ^
     18| 
     19|   const text = readHelloWorld(path);

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

 FAIL  test/hello-world.test.js > can return a value multiple times
AssertionError: expected undefined to be 'hello dir1' // Object.is equality

- Expected: 
"hello dir1"

+ Received: 
undefined

 ❯ test/hello-world.test.js:34:46
     32|   );
     33| 
     34|   expect(readHelloWorld('/tmp/dir1/hw.txt')).toBe('hello dir1');
       |                                              ^
     35|   expect(readHelloWorld('/tmp/dir2/hw.txt')).toBe('hello dir2');
     36| });

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/2]⎯

 Test Files  1 failed (1)
      Tests  2 failed (2)
   Start at  08:38:59
   Duration  755ms (transform 68ms, setup 0ms, collect 84ms, tests 29ms, environment 0ms, prepare 155ms)
```

### System Info

```shell
System:
    OS: macOS 14.5
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
    Memory: 22.86 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.15.1 - ~/.nvm/versions/node/v20.15.1/bin/node
    npm: 10.7.0 - ~/.nvm/versions/node/v20.15.1/bin/npm
  Browsers:
    Chrome: 126.0.6478.183
    Safari: 17.5
  npmPackages:
    @vitest/ui: latest => 2.0.3 
    vite: latest => 5.3.4 
    vitest: latest => 2.0.3
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/docs/guide/mocking.md
+++ b/docs/guide/mocking.md
@@ -383,11 +383,11 @@ module.exports = fs.promises
 :::
 
 ```ts
-// hello-world.js
+// read-hello-world.js
 import { readFileSync } from 'node:fs'
 
 export function readHelloWorld(path) {
-  return readFileSync('./hello-world.txt')
+  return readFileSync(path)
 }
 ```
 
@@ -395,7 +395,7 @@ export function readHelloWorld(path) {
 // hello-world.test.js
 import { beforeEach, expect, it, vi } from 'vitest'
 import { fs, vol } from 'memfs'
-import { readHelloWorld } from './hello-world.js'
+import { readHelloWorld } from './read-hello-world.js'
 
 // tell vitest to use fs mock from __mocks__ folder
 // this can be done in a setup file if fs should always be mocked
@@ -408,7 +408,7 @@ beforeEach(() => {
 })
 
 it('should return correct text', () => {
-  const path = './hello-world.txt'
+  const path = '/hello-world.txt'
   fs.writeFileSync(path, 'hello world')
 
   const text = readHelloWorld(path)
@@ -423,7 +423,7 @@ it('can return a value multiple times', () => {
       './dir2/hw.txt': 'hello dir2',
     },
     // default cwd
-    '/tmp'
+    '/tmp',
   )
 
   expect(readHelloWorld('/tmp/dir1/hw.txt')).toBe('hello dir1')

--- a/packages/vitest/src/runtime/mocker.ts
+++ b/packages/vitest/src/runtime/mocker.ts
@@ -188,7 +188,7 @@ export class VitestMocker {
     return {
       id,
       fsPath,
-      external,
+      external: external ? this.normalizePath(external) : external,
     }
   }
 
@@ -315,14 +315,28 @@ export class VitestMocker {
       const files = readdirSync(mockFolder)
       const baseOriginal = basename(path)
 
-      for (const file of files) {
-        const baseFile = basename(file, extname(file))
-        if (baseFile === baseOriginal) {
-          return resolve(mockFolder, file)
+      function findFile(files: string[], baseOriginal: string): string | null {
+        for (const file of files) {
+          const baseFile = basename(file, extname(file))
+          if (baseFile === baseOriginal) {
+            const path = resolve(mockFolder, file)
+            // if the same name, return the file
+            if (fs.statSync(path).isFile()) {
+              return path
+            }
+            else {
+              // find folder/index.{js,ts}
+              const indexFile = findFile(readdirSync(path), 'index')
+              if (indexFile) {
+                return indexFile
+              }
+            }
+          }
         }
+        return null
       }
 
-      return null
+      return findFile(files, baseOriginal)
     }
 
     const dir = dirname(path)
@@ -517,7 +531,7 @@ export class VitestMocker {
     const mocks = this.mockMap.get(suitefile) || {}
     const resolves = this.resolveCache.get(suitefile) || {}
 
-    mocks[id] = factory || this.resolveMockPath(path, external)
+    mocks[id] = factory || this.resolveMockPath(id, external)
     resolves[id] = originalId
 
     this.mockMap.set(suitefile, mocks)
@@ -546,7 +560,7 @@ export class VitestMocker {
     let mock = this.getDependencyMock(normalizedId)
 
     if (mock === undefined) {
-      mock = this.resolveMockPath(fsPath, external)
+      mock = this.resolveMockPath(normalizedId, external)
     }
 
     if (mock === null) {

--- a/test/core/__mocks__/fs.cjs
+++ b/test/core/__mocks__/fs.cjs
@@ -1,0 +1,6 @@
+// we can also use `import`, but then
+// every export should be explicitly defined
+
+const { fs } = require('memfs')
+
+module.exports = fs

--- a/test/core/__mocks__/fs/promises.cjs
+++ b/test/core/__mocks__/fs/promises.cjs
@@ -1,0 +1,6 @@
+// we can also use `import`, but then
+// every export should be explicitly defined
+
+const { fs } = require('memfs')
+
+module.exports = fs.promises

--- a/test/core/package.json
+++ b/test/core/package.json
@@ -23,6 +23,7 @@
     "axios": "^0.26.1",
     "debug": "^4.3.4",
     "immutable": "5.0.0-beta.5",
+    "memfs": "^4.8.2",
     "strip-ansi": "^7.1.0",
     "sweetalert2": "^11.6.16",
     "tinyrainbow": "^1.2.0",

--- a/test/core/src/read-hello-world.ts
+++ b/test/core/src/read-hello-world.ts
@@ -1,0 +1,6 @@
+// hello-world.js
+import { readFileSync } from 'node:fs'
+
+export function readHelloWorld(path: string) {
+  return readFileSync(path, 'utf-8')
+}

--- a/test/core/test/file-path.test.ts
+++ b/test/core/test/file-path.test.ts
@@ -2,7 +2,11 @@ import { existsSync } from 'node:fs'
 import { describe, expect, it, vi } from 'vitest'
 import { isWindows, slash, toFilePath } from '../../../packages/vite-node/src/utils'
 
-vi.mock('fs')
+vi.mock('fs', () => {
+  return {
+    existsSync: vi.fn(),
+  }
+})
 
 describe('current url', () => {
   it('__filename is equal to import.meta.url', () => {

--- a/test/core/test/mock-fs.test.ts
+++ b/test/core/test/mock-fs.test.ts
@@ -1,0 +1,37 @@
+// hello-world.test.js
+import { beforeEach, expect, it, vi } from 'vitest'
+import { fs, vol } from 'memfs'
+import { readHelloWorld } from '../src/read-hello-world'
+
+// tell vitest to use fs mock from __mocks__ folder
+// this can be done in a setup file if fs should always be mocked
+vi.mock('node:fs')
+vi.mock('node:fs/promises')
+
+beforeEach(() => {
+  // reset the state of in-memory fs
+  vol.reset()
+})
+
+it('should return correct text', () => {
+  const path = '/hello-world.txt'
+  fs.writeFileSync(path, 'hello world')
+
+  const text = readHelloWorld(path)
+  expect(text).toBe('hello world')
+})
+
+it('can return a value multiple times', () => {
+  // you can use vol.fromJSON to define several files
+  vol.fromJSON(
+    {
+      './dir1/hw.txt': 'hello dir1',
+      './dir2/hw.txt': 'hello dir2',
+    },
+    // default cwd
+    '/tmp',
+  )
+
+  expect(readHelloWorld('/tmp/dir1/hw.txt')).toBe('hello dir1')
+  expect(readHelloWorld('/tmp/dir2/hw.txt')).toBe('hello dir2')
+})

